src/gennylib/src/PoolManager.cpp

```diff
@@ -0,0 +1,79 @@
+// Copyright 2019-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gennylib/PoolManager.hpp>
+#include <gennylib/context.hpp>
+#include <gennylib/v1/PoolFactory.hpp>
+
+namespace {
+
+auto createPool(const std::string& mongoUri,
+                genny::PoolManager::CallMeMaybe& apmCallback,
+                genny::WorkloadContext& context) {
+    auto poolFactory = genny::v1::PoolFactory(mongoUri, apmCallback);
+
+    auto queryOpts =
+        context.get_noinherit<std::map<std::string, std::string>, false>("Pool", "QueryOptions");
+    if (queryOpts) {
+        poolFactory.setOptions(genny::v1::PoolFactory::kQueryOption, *queryOpts);
+    }
+
+    auto accessOpts =
+        context.get_noinherit<std::map<std::string, std::string>, false>("Pool", "AccessOptions");
+    if (accessOpts) {
+        poolFactory.setOptions(genny::v1::PoolFactory::kAccessOption, *accessOpts);
+    }
+
+    return poolFactory.makePool();
+}
+
+}  // namespace
+
+
+mongocxx::pool::entry genny::PoolManager::client(const std::string& name,
+                                                 unsigned long instance,
+                                                 genny::WorkloadContext& context) {
+    // Only one thread can access pools.operator[] at a time...
+    this->_poolsGet.lock();
+    LockAndPools& lap = this->_pools[name];
+    // ...but no need to keep the lock open past this.
+    // Two threads trying access client("foo",0) at the same
+    // time will subsequently block on the unique_lock.
+    this->_poolsGet.unlock();
+
+    // only one thread can access the vector of pools at once
+    std::unique_lock lock{lap.first};
+
+    Pools& pools = lap.second;
+
+    while (pools.empty() || pools.size() - 1 < instance) {
+        pools.push_back(createPool(this->_mongoUri, this->_apmCallback, context));
```

Comment:
I agree the reserve & nullcheck is cleaner but I feel icky having "holes" in the vector and I also don't like the possibility of some of the values being nullptr; it's more natural (for me) to reason that any valid entry in the vector `[0..size-1]` being a valid pool / not a nullptr.
I need to think on this for a bit

Reply:
Played with it a bit and I think the "correct" solution here is to use a map rather than a vector. We essentially want a pool associated to each `instance` and there's no real reason why we need pools `[0..n-1]` to exist if we request pool 0. And if we're using `ActorId` as the instance we could end up with `instance` being in the hundreds/thousands so no need to reserve an array(vector) of that size. Big-o stays constant etc.